nexus/src/app/oximeter.rs

```diff
@@ -113,22 +113,55 @@ impl super::Nexus {
         opctx: &OpContext,
         producer_info: nexus::ProducerEndpoint,
     ) -> Result<(), Error> {
-        let (collector, id) = self.next_collector(opctx).await?;
-        let db_info = db::model::ProducerEndpoint::new(&producer_info, id);
-        self.db_datastore.producer_endpoint_create(opctx, &db_info).await?;
-        collector
-            .producers_post(&oximeter_client::types::ProducerEndpoint::from(
-                &producer_info,
-            ))
-            .await
-            .map_err(Error::from)?;
-        info!(
-            self.log,
-            "assigned collector to new producer";
-            "producer_id" => ?producer_info.id,
-            "collector_id" => ?id,
-        );
-        Ok(())
+        for attempt in 0.. {
```

Comment:
I think this loop is an OK compromise, but we could do this all differently. It's a bigger change, but would it help to rewrite the assignment process as a single query? E.g., we could use a subquery to select a random collector, and insert the producer endpoint record with that ID?

Reply:
Yeah, that would definitely be cleaner. I can take a stab at that as a followup? It's a little tricky with the added caveat of "if we're updating instead of inserting, don't do that", I think, but shouldn't be too bad.

Reply:
Follow-up sounds good, however you want to do it. Not pressing, this bit here seems correct and pretty safe.